### PR TITLE
#10360 New sign feature in parameterexpression.py

### PR DESCRIPTION
**Pull request by SamD-1998**
Added a new feature in parameterexpression.py to display the sign of an expression

…the expression inputted

### Summary

I've added a feature that allows a user to display the sign of an inputted expression using an object of the Parameter class.

### CODE REVIEW

**Comment by wshanks** [comment]
Closes #10360

**Review by wshanks — COMMENTED** [review]
Besides in-line comments, please make sure you follow the bot's comment above about the CLA and add a release note as described [here](https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#release-notes).

It would also be good to have a test in test_parameters.py that uses `sign`, though looking now I don't see a test that uses `exp` or `log` there either.

File: qiskit/circuit/parameterexpression.py
Comment (SamD-1998) [reply qiskit/circuit/parameterexpression.py:458]: So I did not do `self._call(func)` on purpose because the issue requires the return type to be numeric. Other methods like `log` would return a `ParameterExpression` but I need a numeric.

**Comment by SamD-1998** [comment]
> 

I don't see a file `test_parameters.py`. I want to show it because I want to explain how a user needs to pass an argument during instantiation for this to work.

**Comment by wshanks** [comment]
> I don't see a file test_parameters.py. I want to show it because I want to explain how a user needs to pass an argument during instantiation for this to work.

The file is [here](https://github.com/Qiskit/qiskit-terra/blob/main/test/python/circuit/test_parameters.py).

File: qiskit/circuit/parameterexpression.py
Comment (wshanks) [reply qiskit/circuit/parameterexpression.py:458]: Why do you need a numeric return type? If the parameter expression is a symbol, don't you need the sign of it to also be a parameter expression until the symbol is assigned a numeric value?

**Comment by SamD-1998** [comment]
Hi, I made a few changes and created a PR again. I forgot how to do it here directly but I think PR initiated it somewhere else. Here is the link to it: 

https://github.com/Qiskit/qiskit-terra/pull/10669

**Comment by Raghav-Bell** [comment]
`linting` tests are failing. Please check tests using `tox` locally. For more please check [Contribution guide](https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#style-and-lint)

File: test/python/circuit/test_parameters.py
Comment (wshanks) [inline test/python/circuit/test_parameters.py]: It's best not to use `print` in a test. Usually, tests contain some form of `self.assert...` to assert some condition that should be true if the code is working correctly. Here, I think you could something like `self.assertEqual(sign_of_parameter.assign(b, -3), -1)`.

File: releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
Comment (wshanks) [inline releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml]:
```suggestion
```

The prelude is used for an overview of the entire release. I don't think we need to call this out there. It is enough to list it in the `features` section.

File: releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
Comment (wshanks) [inline releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml]:
```suggestion
```

This PR does address an open issue, but that issue is more a feature request than a bug, so I don't think we need to list this change in the bug fix section. Listing it in the features section is enough.

File: releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
Comment (wshanks) [inline releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml]:
```suggestion
    Introduced a new feature that adds support for calculating the sign of the Parameter of the value of an expression of a Parameter class.
    Instead of using numpy or other libraries, the user can use the instance of the ParameterExpression class to calculate the sign.
```

**Review by wshanks — CHANGES_REQUESTED** [review]
This is looking close to ready. Thanks for your work, @SamD-1998!

**Comment by wshanks** [comment]
Also, like Raghav noted, the tests are failing because you need to run `black` on the code. There is some information in the guide Raghave linked to. Let us know if you need more help with that part.

**Comment by SameerD-phys** [comment]
Hey, I ran black and I think the formatting issues are now fixed. Just a couple questions:
1) I pushed my commits to my forked repo. I don't want to duplicate a PR again. What should I do?
2) I've signed the CLA but I'm not sure what the second part about adding my email address to this commit to my account is though. Could you please explain?

**Comment by wshanks** [comment]
1. The ideal workflow would be that you 1. checkout the `negative_feature_10360` branch from your fork, 2. run `black`, 3. commit the changes from `black` to your local `negative_feature_10360` branch, and 4. push your local `negative_feature_10360` back to your fork. Maybe you have done the first three steps here and just need to push back to the right branch on your fork?
2. When you create a commit with git, it writes your name and email address into the commit. For your commits made locally, the email address shows up as `sameerdambal@pop-os.localdomain` (probably the default for Pop!OS based on your username). For the commits from you merging in my suggestions, there is a gmail address (I won't copy it here just to avoid inviting spam). I think that gmail address is what is in your GitHub profile and is what is shared with the CLA assistant when you sign it. So the CLA assistant is still unhappy about the pop-os address. You might be able to add that (not real) email address to your GitHub profile to satisfy the CLA assistant. The other option is to use `git config user.email <your email address>` to set your correct email address it git. Then you could do `git rebase -i origin/main` to rebase your changes and mark each commit with `e` to edit it and for each commit then run `git commit --amend --reset-author` and `git rebase --continue` to update the author information. Another option would be to squash all the commits down to one since we will squash them when merging the PR any way. Let me know if you need more help with this part. I think there are several ways to update the author information in git and a lot of articles will come up if you search for it. It's a matter of picking which method seems easiest to follow for you.

File: qiskit/circuit/parameterexpression.py
Comment (wshanks) [inline qiskit/circuit/parameterexpression.py]:
```suggestion
        """Sign of a ParameterExpression"""
```

File: test/python/circuit/test_parameters.py
Comment (wshanks) [inline test/python/circuit/test_parameters.py:958]: For completeness, can we add tests for the other possible results?

```suggestion
        self.assertEqual(sign_of_parameter.assign(b, -3), -1)
        self.assertEqual(sign_of_parameter.assign(b, 2), 1)
        self.assertEqual(sign_of_parameter.assign(b, 0), 0)
```

File: releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
Comment (wshanks) [inline releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml]:
```suggestion
    It can be used as follows::
```

File: releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
Comment (wshanks) [inline releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml]:
```suggestion
    Refer to ` #10360 <https://github.com/Qiskit/qiskit-terra/issues/10360>`__ for more details.
```

File: releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
Comment (wshanks) [inline releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml]:
```suggestion
      from qiskit.circuit import Parameter

      b = Parameter("phi")
      sign_value = b.sign()
      print("sign of Parameter is: ", sign_value)
```

File: releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
Comment (wshanks) [inline releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml]:
```suggestion
    Added support for expressing the sign of a :class:`ParameterExpression`.
```

File: releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
Comment (wshanks) [inline releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml]:
```suggestion
    Instead of assigning a concrete value and using ``numpy.sign`` or other library functions, the user can use the instance of the ParameterExpression class to calculate the sign and can work with the sign before the expression is fully assigned.
```

File: releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
Comment (wshanks) [reply releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml]: I wonder if you want to say something like:

```
print("Sign of an unassigned Parameter is:", sign_value)
print("Sign of a Parameter assigned to -3 is:", sign_value.assign(b, -3))
```

**Review by wshanks — CHANGES_REQUESTED** [review]
I think this is pretty ready to merge! We just need to clean up the release note a bit and deal the CLA issue as discussed previously.

I put some suggestions in for the release note. Mostly these are pointing out syntax issues. The release notes get rendered with Sphinx, so they should follow Sphinx syntax. See [here](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html) for a Sphinx reference.

It is a little tricky to test if the release notes syntax is correct. You need to run sphinx to build the docs to look at. In the Azure Pipelines docs run here, there is an artifact that you can download with the docs from the CI run as well, but currently the release notes are missing (until https://github.com/Qiskit/qiskit/pull/10656 is merged).

**Comment by SameerD-phys** [comment]
Reviewed all recommended changes and pushed a new commit while changing the author information to resolve the CLA issue. I hope it all works out now. Thank you for walking me through all of this!

File: releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
Comment (wshanks) [inline releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml]:
```suggestion
    Added support for expressing the sign of a :class:`.ParameterExpression`
```

File: releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
Comment (wshanks) [inline releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml]:
```suggestion
    Instead of assigning a concrete value and using :data:`numpy.sign` or other library functions, the user can use the instance of the :class:`.ParameterExpression`
```

File: releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
Comment (wshanks) [inline releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml]:
```suggestion
    Refer to `#10360 <https://github.com/Qiskit/qiskit-terra/issues/10360>`__ for more details.
```

**Review by wshanks — CHANGES_REQUESTED** [review]
This PR looks almost ready. Something funny might have happened with your editor. Some of the `` ` `` that I suggested were `'` in your changes. The `` ` `` is a special character used by Sphinx. I put some inline suggestions to switch them to what Sphinx needs.

Your name and email looked good on your most recent commit, but that does not rewrite the history for the older commits GitHub is complaining about. We still need to squash down / rewrite the old commits to update the author information. I can do this at the end if you want, but I think part of the goal with this PR is to help your learn the processes around submitting PR's, so I am happy to keep helping you do this on your side.

**Comment by SameerD-phys** [comment]
Thank you for giving me chance to do this on my own and to learn!
I squashed all commits (or most) and pushed them. I'm not sure if I did it right. Had several issues since `git fetch` would pop up a lot of merge conflicts (most of which did not have a conflict marker and so i did `git rebase --skip` for them). I assume there have been other contributors to this branch and so I need to `git fetch` and resolve merge conflicts everytime.

Could you please identify the errors in the squash commits that I pushed and point them out to me?

**Comment by wshanks** [comment]
Sorry, @SamD-1998, what you did here is not a viable way to fix the CLA issue. Yet, somehow the CLA check has a green checkmark which I don't understand. Did you add the pop-os.localdomain email address to your GitHub account?

I am not sure precisely what you did, but I think you rebased with changes from the main branch merged in which cause those commits to be rewritten and now appear as new commits. This makes it look now like this PR is changing many extra files and has extra authors.

I recommend we roll back and try again. I think the commit before was 981c4707347ae2269241485f06dc3abbc5fb5f47, so the next steps would be:

```
git checkout negative_feature_10360
git reset --hard 981c4707347ae2269241485f06dc3abbc5fb5f47
git push -f origin negative_feature_10360
```

if you are calling this branch `negative_feature_10360` and are calling your fork `origin` in your local git repository.

Sorry about the confusion. I forgot how messy a rebase was after a large branch has been merged in. I think the simplest solution might be to follow the steps given [here](https://stackoverflow.com/a/17141512). In our case, after doing my previous suggestion, this would be something like

```
git checkout negative_feature_10360
git branch -m negfeat-bak
git checkout main
git pull upstream main
git checkout -b negative_feature_10360
git merge --squash negfeat-bak
git commit
git push -f origin negative_feature_10360
```

Let me know if that works for you. If will squash all the previous work down to one commit which you can write a commit message for. When the pull request is approved and merged, it will be squashed to a single commit any way.

**Comment by SameerD-phys** [comment]
Oh that looks essentially like rebasing but manually creating a new branch and merging with the older branch. That's clever. 

Funnily enough, I'm not sure if I added the pop-os.localdomain to my GitHub at this point since I read a lot of things, tried a lot of things and don't remember all the history. I hope this resolves all the issues.

**Comment by wshanks** [comment]
The current diff looks good to me. We have your changes with no extra changes. Some of my last suggestions for the release notes formatting have not been applied, but they still show up in GitHub as in-line comments, so please look at those.

Also, I still see the pop-os.localdomain email in the first commit c1cd914bb7506bac0bde1ba8d19663fa7773e455. It looks like that is a left over initial commit that didn't get squashed. All the changes we want are squashed together in the most recent commit. Maybe you can try the `merge --squash` procedure one more time to get rid of that commit?

**Comment by SameerD-phys** [comment]
Can I ask for some help?

The first commit c1cd914 was unaffected because I missed changing its `pick` to `squash` in the interactive rebase. But as soon as I change it and close the editor, I'm inundated with merge conflicts. I also find myself doing `git rebase --skip` numerous times to ignore these. 

My plan is this:
change `pick` to `squash`
`git add .`
Make formatting changes in releasenotes and then 
```
git checkout negative_feature_10360
git commit
git branch -m negfeat-bak
git checkout main
git pull upstream main
git checkout -b negative_feature_10360
git merge --squash negfeat-bak
git commit
git push -f origin negative_feature_10360
```
This order because I had accommodated your formatting suggestions multiple times but the `git pull upstream main` might have undone it.

Could you please tell me how to resolve this?

**Comment by wshanks** [comment]
I would leave the first commit as `pick` or change it to `reword` and squash the other commits into it. With `reword` or with `git commit --amend` afterwards, you can change the commit message if you want.

Did all the commands run without error for the `git merge --squash` steps I had suggested? That should have left you with a single commit as well. It required that you had defined `upstream` as a remote pointing at `git@github.com:Qiskit/qiskit.git`.

**Comment by wshanks** [comment]
Hmm, you pushed back to an older version without the commits squashed (but still the same final content).

Another option you could try (requires `upstream` pointing at `git@github.com:Qiskit/qiskit.git`) is:

```
git checkout negative_feature_10360
git fetch upstream
git diff --patch upstream/main... > 10360.patch
git reset --hard upstream/main
git apply 10360.patch
git add releasenotes/notes/added-sign-feature-parameterexpression-57693dd69103dc8c.yaml
git commit -a
```

This avoids rebasing and trying rewrite history and just writes your changes to a patch file, resets your branch to upstream/main, and then dumps you changes back out for you to commit as a new clean commit.

**Comment by SameerD-phys** [comment]
Yes, all commands ran without errors for `git merge  --squash` and I think it did leave with a single commit when I checked in `git log`. (Well there was a new commit but I did not know how to go in and check if it had previous commits squashed). I'm surprised that the commits did not appear to be squashed on your end. 

My `upstream` pointed to `https://github.com/Qiskit/qiskit-terra.git`. earlier. Now I've made it point to `git@github.com:Qiskit/qiskit.git`.

The patch file  method did not work because the patch file was blank. I read that this indicates that there is no difference between those branches.

I feel rebasing is quite hard with a large number of commits present which give a lot of conflicts while interactive rebasing.

**Comment by wshanks** [comment]
> The patch file method did not work because the patch file was blank. I read that this indicates that there is no difference between those branches.

Hmm, there are differences, as long as you didn't reset your local branch to upstream/main.

If you want, you can try generating the patch again and try to figure out why it doesn't show your changes.

If you want to skip the `git diff` part, I am attaching the output patch file that I get with this message.
[10360.patch.txt](https://github.com/Qiskit/qiskit/files/12588744/10360.patch.txt)

**Comment by SameerD-phys** [comment]
Hi, I ran these commands. I'm not sure if it worked this time around although I've been trying to push the releasenotes with the suggested changes for quite some time now.

If the issue still persists, is there a possibility of resolving it on your end?

**Comment by wshanks** [comment]
Okay, I force pushed a commit with the patch that I shared above with your information for the commit author that matches what the CLA check accpets. This restores the state to the last state you had pushed changes to. I still had some suggestions in-line for the release notes that are valid. Can you address them? You can just accept and commit my suggestions if you want.

**Review by wshanks — APPROVED** [review]
All good now, thanks for working on this!

**Review by jakelishman — APPROVED** [review]
(proxying Will's approval, since the PR touches files Will's not codeowner on)

**Pull request merged**
